gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/DatasetBase.java

```diff
@@ -0,0 +1,215 @@
+ *
+ * <p>
+ *   Concrete subclasses should implement {@link #getVersionFinder} and {@link #getRetentionPolicy}.
+ * </p>
+ *
+ * <p>
+ * Datasets are directories in the filesystem containing data files organized in version-like directory structures.
+ * Example datasets:
+ * </p>
+ *
+ * <p>
+ *   For snapshot based datasets, with the directory structure:
+ *   <pre>
+ *   /path/to/table/
+ *      snapshot1/
+ *          dataFiles...
+ *      snapshot2/
+ *          dataFiles...
+ *   </pre>
+ *   each of snapshot1 and snapshot2 are dataset versions.
+ * </p>
+ *
+ * <p>
+ *   For tracking datasets, with the directory structure:
+ *   <pre>
+ *   /path/to/tracking/data/
+ *      2015/
+ *          06/
+ *              01/
+ *                  dataFiles...
+ *              02/
+ *                  dataFiles...
+ *   </pre>
+ *   each of 2015/06/01 and 2015/06/02 are dataset versions.
+ * </p>
+ *
+ * <p>
+ *  {@link DatasetBase} uses a {@link gobblin.data.management.retention.version.finder.DatasetVersionFinder} to find all
+ *  subdirectories that are versions of this dataset. After that, for each dataset, it uses a
+ *  {@link gobblin.data.management.retention.policy.RetentionPolicy} to decide which versions of the dataset should be
+ *  deleted. For each version deleted, if {@link #deleteEmptyDirectories} it will also look at all parent directories
+ *  and delete directories that are now empty, up to but not including the dataset root.
+ * </p>
+ *
+ * @param <T> type of {@link gobblin.data.management.retention.version.DatasetVersion} supported by this
+ *           {@link gobblin.data.management.retention.dataset.Dataset}.
+ */
+public abstract class DatasetBase<T extends DatasetVersion> implements Dataset {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatasetBase.class);
+
+  public static final String CONFIGURATION_KEY_PREFIX = "gobblin.retention.";
+  public static final String SIMULATE_KEY = CONFIGURATION_KEY_PREFIX + "simulate";
+  public static final boolean SIMULATE_DEFAULT = false;
+  public static final String SKIP_TRASH_KEY = CONFIGURATION_KEY_PREFIX + "skip.trash";
+  public static final boolean SKIP_TRASH_DEFAULT = false;
+  public static final String DELETE_EMPTY_DIRECTORIES_KEY = CONFIGURATION_KEY_PREFIX + "delete.empty.directories";
+  public static final boolean DELETE_EMPTY_DIRECTORIES_DEFAULT = true;
+
+  protected final FileSystem fs;
+  protected final Trash trash;
+  protected final boolean simulate;
+  protected final boolean skipTrash;
+  protected final boolean deleteEmptyDirectories;
+
+  /**
+   * Get {@link gobblin.data.management.retention.version.finder.VersionFinder} to use.
+   */
+  public abstract VersionFinder<? extends T> getVersionFinder();
+
+  /**
+   * Get {@link gobblin.data.management.retention.policy.RetentionPolicy} to use.
+   * @return
+   */
+  public abstract RetentionPolicy<T> getRetentionPolicy();
+
+  public DatasetBase(FileSystem fs, Props props) throws IOException {
```

Comment:
This constructor and the DatasetBase(FileSystem fs, Trash trash, boolean simulate, boolean skipTrash, boolean deleteEmptyDirectories) constructor should be sharing code or the former should be using the latter. Otherwise, we are risking having inconsistent logic in the constructor.